chain-abci/src/enclave_bridge/edp/tdbe.rs

```diff
@@ -0,0 +1,238 @@
+use std::{
+    collections::HashMap,
+    future::Future,
+    io::{self, Cursor, Seek, SeekFrom},
+    os::unix::net::UnixStream,
+    pin::Pin,
+    sync::Arc,
+    thread,
+};
+
+use aesm_client::AesmClient;
+use enclave_runner::{
+    usercalls::{AsyncListener, AsyncStream, UsercallExtension},
+    EnclaveBuilder,
+};
+use kvdb::KeyValueDB;
+use serde::{Deserialize, Serialize};
+use sgxs_loaders::isgx::Device;
+use tdbe_common::TdbeStartupConfig;
+use tokio::net::{TcpListener, TcpStream};
+
+use chain_core::tx::data::TxId;
+use chain_storage::ReadOnlyStorage;
+use enclave_protocol::{
+    codec::{StreamRead, StreamWrite},
+    tdbe_protocol::PersistenceCommand,
+    EnclaveRequest, EnclaveResponse, SealedLog,
+};
+use ra_sp_server::config::SpRaConfig;
+
+#[derive(Serialize, Deserialize, Debug, Default)]
+pub struct TdbeConfig {
+    /// Optional TM RPC address of another TDBE server from where to fetch data
+    pub remote_rpc_address: Option<String>,
+    /// Local TDBE server address to listen on. E.g. `127.0.0.1:3445`
+    pub local_listen_address: String,
+    /// External TDBE server address, used by remote nodes to send RPC requests. E.g.
+    /// `<public_ip>:<public_port>`
+    pub external_listen_address: String,
+}
+
+#[derive(Debug)]
+pub struct TdbeApp {
+    /// UDS to connect to `chain-abci`
+    chain_abci_stream: UnixStream,
+    /// UDS to persist data to `chain-storage`
+    persistence_stream: UnixStream,
+    /// `ra-sp-server` address for remote attestation. E.g. `0.0.0.0:8989`
+    /// TODO:  Replace it with a local UDS (using `chain-abci` as launcher).
+    sp_address: String,
+    /// Optional address of remote TDBE server for fetching initial data
+    remote_tdbe_address: Option<String>,
+    /// Optional DNS name of remote TDBE server for fetching initial data (`chain-abci` fetches it
+    /// from RPC and passes it to TDBE App)
+    remote_tdbe_dns_name: Option<String>,
+    /// Local TDBE server address to listen on. E.g. `127.0.0.1:3445`
+    local_listen_address: String,
+}
+
+impl TdbeApp {
+    /// Creates a new instance of TDBE app
+    pub fn new(
+        tdbe_config: &TdbeConfig,
+        ra_config: &SpRaConfig,
+        storage: Arc<dyn KeyValueDB>,
+    ) -> std::io::Result<Self> {
+        let (chain_abci_stream, chain_abci_receiver) = UnixStream::pair()?;
+        let (persistence_stream, persistence_receiver) = UnixStream::pair()?;
```

Comment:
how adding comment about?
who is server, and who is client

Reply:
Added comments